addons/binding/org.openhab.binding.modbus/DEVELOPERS.md

```diff
@@ -0,0 +1,111 @@
+2. Download [diagslave](http://www.modbusdriver.com/diagslave.html) and start modbus serial slave up using this command:
+
+```
+./diagslave -m rtu -a 1 -b 38400 -d 8 -s 1 -p none -4 10 /dev/pts/7
+```
+
+3. Configure openHAB's modbus slave to connect to `/dev/pts/8`:
+
+```
+xxx.connection=/dev/pts/8:38400:8:none:1:rtu
+```
+
+4. Modify `start.sh` or `start_debug.sh` to include the unconventional port name by adding the following argument to `java`:
+
+```
+-Dgnu.io.rxtx.SerialPorts=/dev/pts/8
+```
+
+Naturally this is not the same thing as the real thing but helps to identify simple issues.
+
+## Testing TCP Implementation
+
+1. Download [diagslave](http://www.modbusdriver.com/diagslave.html) and start modbus tcp server (slave) using this command:
+
+```
+./diagslave -m tcp -a 1 -p 55502
+```
+
+2. Configure openHAB's modbus slave to connect to `127.0.0.1:55502`:
+
+```
+tcp.slave1.connection=127.0.0.1:55502
+```
+
+## Writing Data
+
+See this [community post](https://community.openhab.org/t/something-is-rounding-my-float-values-in-sitemap/13704/32?u=ssalonen) explaining how `pollmb` and `diagslave` can be used to debug modbus communication.
+
+You can also use `modpoll` to write data:
+
+
+```bash
+# write value=5 to holding register 40001 (index=0 in the binding)
+./modpoll -m tcp -a 1 -r 1 -t4 -p 502 127.0.0.1 5
+# set coil 00001 (index=0 in the binding) to TRUE
+./modpoll -m tcp -a 1 -r 1 -t0 -p 502 127.0.0.1 1
+# write float32
+./modpoll -m tcp -a 1 -r 1 -t4:float -p 502 127.0.0.1 3.14
+```
+
+## Extending Modbus binding
+
+This Modbus binding can be extended by other OSGi bundles to add more specific support for Modbus enabled devices. To do so to you have to create a new OSGi bundle which has the same binding id as this binding. The best way is to use the `ModbusBindingConstants.BINDING_ID` constant.
+
+You'll have to create one or more handler classes for the devices you want to support. For the modbus connection setup and handling you can use the Modbus TCP Slave or Modbus Serial Slave handlers. Your handler should use these handlers as bridges and you can set up your regular or one shot modbus requests to read from the slave. This is done by by creating a `BasicPollTaskImpl` and submitting it using the `ModbusManager` `submitOneTimePoll` and `registerRegularPoll` methods.
+
+Please keep in mind that these reads are asynchronous and they will call your callback once the read is done.
+
+Once you have your data read from the modbus device you can parse and transform them then update your channels to publish these data to the openHAB system.
+
+### Discovery
+
+If you write a device specific handler then adding discovery for this device is very welcome. You will have to write a discovery participant class which implements the `ModbusDiscoveryParticipant` interface and registers itself as a component. Example:
+
+```java
+
+@Component(immediate = true)
+public class SunspecDiscoveryParticipant implements ModbusDiscoveryParticipant {
```

Comment:
[Class should have @NonNullByDefault](https://www.openhab.org/docs/developer/development/conventions.html#null-annotations).

Reply:
done